File: 447_NumberOfBoomerangs.py
# 42.86%
class Solution(object):
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        def distance(point1, point2):
            return (point1[0] - point2[0]) * (point1[0] - point2[0]) + (point1[1] - point2[1]) * (point1[1] - point2[1])
            
        size, res = len(points), 0
        #map = dict{}
        for i in range(size):
            distMap = dict()
            for j in range(size):
                d = distance(points[i], points[j])
                if d in distMap:
                    distMap[d] += 1
                else:
                    distMap[d] = 1
            for value in distMap.values():
                #if value >= 2:
                 res += (value - 1) * value;
        return res

##This one looks like mine		
class Solution(object):
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """	
		res = 0
        for p in points:
            cmap = {}
            for q in points:
                f = p[0]-q[0]
                s = p[1]-q[1]
                cmap[f*f + s*s] = 1 + cmap.get(f*f + s*s, 0)
            for k in cmap:
                res += cmap[k] * (cmap[k] -1)
        return res
        

File: 447_NumberOfBoomerangs.cpp
//Memory limit exceeded
class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int pointNum(points.size());
		int res(0);
		vector<map<int, int> > distPoints(pointNum);
		for(int i = 0; i < pointNum - 1; i++)
			for(int j = i + 1; j < pointNum; j++){
				int d(distance(points[i], points[j]));
				distPoints[i][d]++;
				distPoints[j][d]++;
			}
		for(auto &distCounter : distPoints){
			for(auto &counter : distCounter){
				if(counter.second > 1)
					res += ((counter.second - 1) * counter.second);
			}
		}
		return res;	
	}
private:
	inline int distance(const pair<int, int>& point1, const pair<int, int>& point2){
		int x1(point1.first), y1(point1.second), x2(point2.first), y2(point2.second);
		return (x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2);
	}
};

//88.16%
class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int pointNum(points.size());
		int res(0);
		for(int i = 0; i < pointNum; i++){
			map<int, int> distCounter;
			for(int j = 0; j < pointNum; j++){
				int d(distance(points[i], points[j]));
				distCounter[d]++;
			}
			for(auto& item : distCounter){
				//if(item.second > 1)
				res += ((item.second - 1) * item.second);
			}
		}
		return res;	
	}
private:
	inline int distance(const pair<int, int>& point1, const pair<int, int>& point2){
		int x1(point1.first), y1(point1.second), x2(point2.first), y2(point2.second);
		return (x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2);
	}
};

//Wonderful solution from LeetCode
int numberOfBoomerangs(vector<pair<int, int>>& points) {
    int booms = 0;
    for (auto &p : points) {
        unordered_map<double, int> ctr(points.size());
        for (auto &q : points)
            booms += 2 * ctr[hypot(p.first - q.first, p.second - q.second)]++;
    }
    return booms;
}

